### ISSUE TITLE
Feature Request:  S3 Select Object as JDBC ( or Athena results as S3 Select results)

### ISSUE THREAD

**Issue by DALDEI** [burtcorp/athena-jdbc#18]
Ive been playing with S3 SelectObject and am very impressed.
When given a query that it can do (very limited subset of Athena -- very VERY limited) 
it does so very well  compared to athena.
My use case is a very small (wrt  'big data') set of JSON and CSV files -- about 1000 files that are highly partitioned (by file name not hive partitioning ).  Much of what I could use athena for instead Im reading from S3 directly -- because the latency of even the  simplest athena query is rarely less then 2 seconds.   Until I hit 100+ files I can out-perform athena by far - but I dont like writing all that code :)
I tried out S3 Select Object -- in the *worse* case (select *)  it performs approx the same as s3 get-object.  50-200ms latancy when running off of the same region EC2.
I can read 100 files in parallel in < 1 sec. FOr better cases where either a subset of columns or some aggregation is done its much faster.

But -- I have yet to use it much because its an entirely different result set format -- 
I will probalby end up wrapping some ugly high level class structure that merges JDBC/Athena,  'csv' file reads from S3 get-object and  S3 Select so I can use them some what interchanably.

However this library look like a much cleaner place to do this then on top of it.

Feature Request:  provide the same or compatible interface to run S3 Select as Athena -- mainly the result set.  It doesn't matter to me if thats JDBC ResultSet or something else -- just having fewer versions of a 'streaming CSV-ish record set' would be very useful.

If that were done -- it may well uncover more fun things that are unnecessary :) 
Like using S3 Select on top of athena result sets .. 

Even as  a drop-in replacement to s3 get-object (for CSV or JSON) is a benfit in code simplicity --

**Comment by DALDEI** [comment]
Below is some POC-quality kotlin code Im currently using for basic S3 select 
The hardest part is getting the serialization objects configured.



```
...
import ...



private val log = logger{}
fun <T> logTime(msg: String, block: () -> T ) : T {
  val start = System.nanoTime()
  return block().also {
    val end = System.nanoTime()
      log.debug{"${msg}: elapsed: ${ (end-start) / 1_000_000_000.0 } sec"}
  }
}

object RecordQuery {

  val s3cache = AWSClientCache(AmazonS3ClientBuilder::standard)
  val region: String get() = SystemConfig.aws.region

  fun readObject( docuri: DOCURI ) = when {

    docuri.isScheme( "s3" ) -> logTime("readObjectS3" ) { readObjectS3(docuri,SchemaFormat.JSON ) }
    else ->  logTime("readObject DOCURI.inputStream") {  docuri.inputStream() }
  }
  fun readObjectS3(docuri : DOCURI , asFormat:  SchemaFormat  )   = readObjectS3(docuri, when(asFormat) {
    SchemaFormat.JSON -> FlatSchema.jsonFormat("")
    SchemaFormat.CSV -> FlatSchema.csvFormat("")
  })

  fun readObjectS3(docuri : DOCURI , asFormat : FlatSchema   ) : InputStream  = queryObjectS3(docuri,asFormat,"SELECT s.* from S3Object s")

  fun queryObjectS3(docuri : DOCURI , asFormat : FlatSchema, query: String    ) : InputStream {
    val s3uri = AmazonS3URI(docuri.uri)
    val pn = docuri.parsedName
    return with( s3cache.get(region) ){
      selectObjectContent(  SelectObjectContentRequest().apply {
        bucketName = s3uri.bucket
        key = s3uri.key
        expressionType = "SQL"
        inputSerialization = InputSerialization().apply {
          val schema = pn.cType.flatSchema
          when (schema.defaultFormat ) {
            SchemaFormat.CSV -> {
              val opts = schema.csvOptions
              csv = CSVInput().apply {
                fileHeaderInfo = if (opts.header) "Use" else "None" //Valid values: None, Ignore, Use.
                // quoteEscapeCharacter =
                fieldDelimiter = opts.separator
                setRecordDelimiter(opts.eol)
                allowQuotedRecordDelimiter = opts.quoted
                quoteCharacter = '"'

              }
            }
            SchemaFormat.JSON -> {
              val opts = schema.jsonOptions
              json = JSONInput().apply {
                type = when (opts.jsonFormat) {
                  JsonFormat.JSON_LINES -> "LINES"
                  else -> "DOCUMENT"
                }
              }
            }
          }
        }
        outputSerialization = OutputSerialization().apply {
          val schema = pn.cType.flatSchema
          when (asFormat.defaultFormat) {
            SchemaFormat.JSON -> {
              json = JSONOutput().apply {
                recordDelimiter = '\n'
              }
            }
            SchemaFormat.CSV -> {
              val opts = schema.csvOptions
                csv = CSVOutput().apply {
                  //fileHeaderInfo = if (opts.header) "Use" else "None" //Valid values: None, Ignore, Use.
                  // quoteEscapeCharacter =
                  fieldDelimiter = opts.separator
                  setRecordDelimiter(opts.eol)
                  quoteFields  = if( opts.quoted ) "ALWAYS" else "ASNEEDED"
                  quoteCharacter = '"'

                }
            }
          }
        }
        expression = query

      }

      ).payload.recordsInputStream

    }
  }
}

```

**Comment by grddev** [comment]
Interesting idea, but unfortunately, I think the amount of code shared will be significantly less than the additional complexity. Especially as it seems [S3 Select is not yet supported by aws-sdk-java-v2](https://github.com/aws/aws-sdk-java-v2/issues/859), which is the S3 dependency used in this project.

**Comment by iconara** [comment]
Excellent idea, but as @grddev says, probably out of scope for this project. JDBC is unfortunately a _very_ big spec, and a lot of the work into making this one has been just to implement all the details – and most of that is not transferrable to a different backend.

**Comment by DALDEI** [comment]
Closing issue. 
From a user perspective there is an additional layer of complexity due to the limited subset of SQL that S3 select supports its not even close to a 'drop in replacement' even if the jdbc layer did it.
